### PR TITLE
EL-4253: Dashboard: widgets overlap when dragging down

### PR DESCRIPTION
**Pull request by JonathanToddMF**
#### Checklist

* [x] The contents of this PR are mine to submit. No third party code or other material is being committed.
* [x] New third party dependencies (if any) are linked via `package.json`. Third party dependencies are licensed under one of the following: Apache, MIT, BSD, Mozilla Public License, Eclipse Public License, or Oracle Binary Code License.
* [x] The code in this PR does not contain export-restricted encryption algorithms.

#### Ticket / Issue

https://portal.digitalsafe.net/browse/EL-4253

#### Description of Proposed Changes

1. Changed dashboard service to target left/right movement based on the width of the actionWidget and not colspan of the widget 
2. Changed dashboard service to move the widget the width of the action widget and not just 1 column
3. Added new 8 column dashboard test suit
4. Added test coverage for shifting widgets to the right and left

#### Documentation CI URL

https://pages.github.houston.softwaregrp.net/sepg-docs-qa/UXAspects_CI_UXAspects_EL-4253-dashboard-widgets-overlap

#### Downstream Build(s)

[ux-aspects-micro-focus](https://jenkins.swinfra.net/job/SEPG/job/caf/view/Developer/job/caf~ux-aspects-micro-focus~EL-4253-dashboard-widgets-overlap~CI/)

### CODE REVIEW

**Comment by donna-marijne** [comment]
Can you merge in the changes and make sure that the downstream build passes?

File: src/components/dashboard/dashboard.service.ts [hunk 1]
```diff
@@ -697,7 +697,7 @@ export class DashboardService implements OnDestroy {
 
         // find the positions required
         const targetSpaces = this.getOccupiedSpaces().filter(space => space.widget === widget).map(space => {
-            return { column: space.column - widget.getColumnSpan(), row: space.row, widget: space.widget };
+            return { column: space.column - this._actionWidget.widget.getColumnSpan(), row: space.row, widget: space.widget };
```
Comment (donna-marijne) [inline src/components/dashboard/dashboard.service.ts:700]: I don't think this change is appropriate - see the `if` statement above. The function is intended to work on the parameter provided.

File: src/components/dashboard/dashboard.service.ts [hunk 2]
```diff
@@ -737,7 +737,7 @@ export class DashboardService implements OnDestroy {
 
         // find the positions required
         const targetSpaces = this.getOccupiedSpaces().filter(space => space.widget === widget).map(space => {
-            return { column: space.column + widget.getColumnSpan(), row: space.row, widget: space.widget };
+            return { column: space.column + this._actionWidget.widget.getColumnSpan(), row: space.row, widget: space.widget };
```
Comment (donna-marijne) [inline src/components/dashboard/dashboard.service.ts:740]: Revert this as well.

**Review by donna-marijne — CHANGES_REQUESTED** [review]
Thanks for the PR. Looks like there are some inappropriate changes here. I think the change on line 759 all that's required.

File: src/components/dashboard/dashboard.service.ts (hunk 1 above)
Comment (JonathanToddMF) [reply src/components/dashboard/dashboard.service.ts:700]: Hi Donna, the reason I included this is if there is another widget placed beside the one being moved it causes it to moved right a bit too far. See this video for a look at the issue - https://gyazo.com/cbd43aa31d52dfe933672f2670fbfb5f

From my investigation it seemed targetSpaces is where the widget is planning to move right. Without this, it was being calculated to be columns 4-7 (this is why the alert widget moves further than necessary) but in reality the widget should only move to occupy the width of the action widget ie columns 2-5. I hope this explains the thought process behind this!

Comment (donna-marijne) [reply src/components/dashboard/dashboard.service.ts:700]: That's fair. I noticed this issue, which I assumed was due to this change, but perhaps I misunderstood:
[Image from Gyazo](https://gyazo.com/93a4f21cbe7ffedb72ff14650e31b00c)

**Comment by JonathanToddMF** [comment]
Waiting for builds on this

**Comment by JonathanToddMF** [comment]
Ready for review

**Review by donna-marijne — DISMISSED** [review]
Verification:

https://plnkr.co/edit/3UqY2FqThbOilB28?preview - :heavy_check_mark:

https://plnkr.co/edit/lpkkoYBafxNyfp71?preview - :heavy_check_mark:

**Comment by donna-marijne** [comment]
For future reference, if you have a subcategory of tests such as "flipped," you can nest `describe`s and provide separate `beforeEach`/`beforeAll` hooks.

**Comment by donna-marijne** [comment]
@ashley-hunter Assigning for acceptance purposes as discussed previously.

File: src/components/dashboard/dashboard.service.ts [hunk 3]
```diff
@@ -1412,6 +1467,11 @@ export interface DashboardLayoutData {
     minRowSpan?: number;
 }
 
+export enum HorizontalMovement {
+    Left = -1,
+    Right = 1
+}
```
Comment (ashley-hunter) [inline src/components/dashboard/dashboard.service.ts:1473]: This doesn't seem to ever be used?

File: src/components/dashboard/dashboard.service.ts [hunk 4]
```diff
@@ -706,17 +733,17 @@ export class DashboardService implements OnDestroy {
...
-            targetSpaces.forEach(space => this.getWidgetsAtPosition(space.column, space.row).filter(wgt => wgt !== space.widget).forEach(wgt => this.canWidgetMoveLeft(wgt, true)));
+            // // move all widgets to the left
+            targetSpaces.forEach(space => this.getWidgetsAtPosition(space.column, space.row).filter(wgt => wgt !== space.widget).forEach(wgt => this.canWidgetMoveLeft(wgt, true, colShift)));
 
-            // find the target column
+            // // find the target column
             const column = targetSpaces.reduce((target, space) => Math.min(target, space.column), Infinity);
 
-            // move current widget to the left
+            // // move current widget to the left
```
Comment (ashley-hunter) [inline src/components/dashboard/dashboard.service.ts:746]: These have been double commented?

**Review by ashley-hunter — CHANGES_REQUESTED** [review]
Thanks, looks good and seems to work well, just noticed some unused code I think we can remove

**Comment by JonathanToddMF** [comment]
Thanks Ashley, just waiting on builds before re-requesting review

**Review by ashley-hunter — APPROVED** [review]
Thanks! Looks good to me!

**Pull request merged**